### PR TITLE
 sike mitigation with test cases

### PR DESCRIPTION
**Pull request by zhdllwyc**

### CODE REVIEW

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 1]
```diff
@@ -98,3 +102,207 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+	mul(x, z, &t1)           // X2p = Z2p * t1
+	sub(&t1, &t1, &t0)       // t1  = t1 - t0
+	mul(&t0, &params.A, &t1) // t0  = A24+ * t1
+	add(z, z, &t0)           // Z2p = Z2p + t0
+	mul(z, z, &t1)           // Zp  = Z2p * t1
+}
+
+
+
+// Right-to-left Montgomery ladder, Algorithm 4 in Costello-Smith
```
Comment (armfazh) [inline dh/sidh/internal/templates/curve_test.gotemp:130]:
```suggestion
// Left-to-right Montgomery ladder, Algorithm 4 in Costello-Smith
```

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 2]
```diff
@@ -98,3 +102,207 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+
+	mul(&pub3Pt[0], &P.X, &invPz)
+	mul(&pub3Pt[1], &Q.X, &invQz)
+}
+
+/* -------------------------------------------------------------------------
+   Public key / Ciphertext validation against attacks proposed in paper https://eprint.iacr.org/2022/054.pdf and HB
+   -------------------------------------------------------------------------*/
+
+func TestInvalidPK(t *testing.T) {
```
Comment (armfazh) [inline dh/sidh/internal/templates/curve_test.gotemp:249]: here we can split this large test in three single (sub-)tests.

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 3]
```diff
@@ -98,3 +102,207 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+// Construct Invalid public key tuple (P,Q) such that P and Q are linearly dependent 
+// Simulate section 3.1.1 of paper https://eprint.iacr.org/2022/054.pdf
+// We only construct point P and Q because in the attacks the third point is P-Q by construction 
+// and the countermeasure does not test it
+// Without loss of generality, we assume the curve is the starting curve 
+func InvalidPKNoneLinear(scalar []uint8, pub3Pt *[3]Fp2) {
+	
+	var P, Q ProjectivePoint
+	
+	rand.Seed(time.Now().UnixNano())
```
Comment (armfazh) [inline dh/sidh/internal/templates/curve_test.gotemp:167]: we may want to keep track of the seed that was used, this is in the case that the test fails so we can reproduce it.

**Review by armfazh — CHANGES_REQUESTED** [review]
some minor changes needed

File: dh/sidh/internal/templates/curve.gotemp [hunk 4]
```diff
@@ -365,3 +360,83 @@ func (phi *isogeny4) EvaluatePoint(p *ProjectivePoint) {
...
+	var PZ, QZ Fp2
+	FromMontgomery(&PZ, &test_P.Z)
+	FromMontgomery(&QZ, &test_Q.Z)
+
+	
+	// P, Q are not of full order 3^e_3
+	if((isZero(&PZ)==1)||(isZero(&QZ)==1)){
+		//fmt.Println("Not full order");
+		return errors.New("curve: ciphertext/public key are not of full order 3^e3")
+	}
```
Comment (bwesterb) [inline dh/sidh/internal/templates/curve.gotemp:409]: This check assumes `ord(P) | 3^(e_3)` (and similar for `Q`) — how do we know that?

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 5]
```diff
@@ -98,3 +102,314 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
 	}
 }
+
+/* -------------------------------------------------------------------------
+   Generate invalid public key points / ciphertext for test TestKEMInvalidPK
+   -------------------------------------------------------------------------*/
+
+// Given the curve parameters, xP = x(P), computes xP = x([2]P)
+// Safe to overlap xP, x2P.
+func xDbl(xP *ProjectivePoint, params *CurveCoefficientsEquiv) {
```
Comment (bwesterb) [inline dh/sidh/internal/templates/curve_test.gotemp:112]: Can't you use `Pow2k`?

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 6]
```diff
@@ -98,3 +102,314 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+		cswap(&R0.X, &R0.Z, &R1.X, &R1.Z, swap)
+		R0, R1 = xDbladd(&R0, &R1, &R2, &aPlus2Over4)
+	}
+	cswap(&R0.X, &R0.Z, &R1.X, &R1.Z, prevBit)
+	return R0
+}
+
+// P = P + T
+func tauT(P *ProjectivePoint) {
+	P.X, P.Z = P.Z, P.X	// magic!
```
Comment (bwesterb) [inline dh/sidh/internal/templates/curve_test.gotemp:155]: Explain what T is and why this works.

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 7]
```diff
@@ -98,3 +102,314 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+	}
+
+}
+
+
+/* -------------------------------------------------------------------------
+   Public key / Ciphertext validation against attacks proposed in paper https://eprint.iacr.org/2022/054.pdf and HB
+   -------------------------------------------------------------------------*/
+
+func TestInvalidPK(t *testing.T) {
```
Comment (bwesterb) [inline dh/sidh/internal/templates/curve_test.gotemp:406]: We can have a separate top-level test for each. Then it's easier to see which one failed, if they fail.

**Review by bwesterb — CHANGES_REQUESTED** [review]
Thanks @zhdllwyc, this is great.

File: dh/sidh/internal/p434/cov.txt [hunk 8]
```diff
@@ -0,0 +1,131 @@
+mode: set
+github.com/cloudflare/circl/dh/sidh/internal/p434/arith_generic.go:16.34,20.31 2 1
```
[the remark this answers is not in the document]
Comment (zhdllwyc) [reply dh/sidh/internal/p434/cov.txt:2]: Yes!

File: dh/sidh/internal/templates/core.gotemp [hunk 9]
```diff
@@ -274,6 +276,15 @@ func DeriveSecretB(ss, prv []byte, pub3Pt *[3]Fp2) {
 	xP = ProjectivePoint{X: pub3Pt[0], Z: params.OneFp2}
 	xQ = ProjectivePoint{X: pub3Pt[1], Z: params.OneFp2}
 	xQmP = ProjectivePoint{X: pub3Pt[2], Z: params.OneFp2}
+
+	//PUBLIC KEY VALIDATION
+	if err := PublicKeyValidation(&cparam, &xP, &xQ, &xQmP, params.B.SecretBitLen); err != nil {
+		_, err_read := io.ReadFull(crand.Reader, ss)
+		if err_read != nil {
+			panic("core: failed to generate random ss when public key verification fails")
+		}
```
[the remark this answers is not in the document]
Comment (zhdllwyc) [reply dh/sidh/internal/templates/core.gotemp:285]: yes, thanks

File: dh/sidh/internal/templates/curve.gotemp [hunk 10]
```diff
@@ -365,3 +360,83 @@ func (phi *isogeny4) EvaluatePoint(p *ProjectivePoint) {
...
+		//fmt.Println("pmq is not right");
+		return errors.New("curve: PmQ is invalid")
+	}
+	
+	cparam := CalcCurveParamsEquiv3(cparams)
+
+	// Compute e_3 = log3(2^(nbits+1))
+	var e3 uint32
+	e3_float := float64(int(nbits)+1)/math.Log2(3)
+	e3 = uint32(e3_float)
```
[the remark this answers is not in the document]
Comment (zhdllwyc) [reply dh/sidh/internal/templates/curve.gotemp:388]: Having e3 as a parameter requires me to change the params.go
I want to put changes in the test files as much as possible.
Do you think I should touch the params.go?

File: dh/sidh/internal/templates/curve.gotemp (hunk 4 above)
Comment (zhdllwyc) [reply dh/sidh/internal/templates/curve.gotemp:409]: Yes, if ord(P) does not divide 3^e_3.
Will be caught by:
// P, Q are not of correct order 3^e_3
if((isZero(&PZ)==0)||(isZero(&QZ)==0)){
		return errors.New("curve: ciphertext/public key are not of correct order 3^e3")
}

File: dh/sidh/internal/templates/curve_test.gotemp (hunk 5 above)
Comment (zhdllwyc) [reply dh/sidh/internal/templates/curve_test.gotemp:112]: Yes! Thanks for reminding me this

File: dh/sidh/internal/templates/curve_test.gotemp (hunk 7 above)
Comment (zhdllwyc) [reply dh/sidh/internal/templates/curve_test.gotemp:406]: sure, sounds good

File: dh/sidh/internal/templates/core.gotemp [hunk 11]
```diff
@@ -260,7 +260,7 @@ func DeriveSecretA(ss, prv []byte, pub3Pt *[3]Fp2) {
 }
 
 // Establishing shared keys in in 3-torsion group
-func DeriveSecretB(ss, prv []byte, pub3Pt *[3]Fp2) {
+func DeriveSecretB(ss, prv []byte, pub3Pt *[3]Fp2) int{
```
[the remark this answers is not in the document]
Comment (zhdllwyc) [reply dh/sidh/internal/templates/core.gotemp]: sure, sounds good

File: dh/sidh/internal/templates/curve_test.gotemp (hunk 2 above)
Comment (zhdllwyc) [reply dh/sidh/internal/templates/curve_test.gotemp]: sure, sounds good

File: dh/sidh/internal/templates/curve.gotemp (hunk 10 above)
[the remark this answers is not in the document]
Comment (armfazh) [reply dh/sidh/internal/templates/curve.gotemp:387]: e3 is definitely a constant, however, its calculation is simple and used once.
We may revisit later moving this value to the params files.

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 12]
```diff
@@ -98,3 +102,316 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+
+	testInvalidPmQ(t)
+	testInvalidPKNoneLinear(t)
+	testInvalidPKT(t)
+	testInvalidPKOrder2(t)
+	testInvalidPKFullOrder(t)
+	testValidPQ(t)
```
Comment (armfazh) [inline dh/sidh/internal/templates/curve_test.gotemp:415]:
```suggestion
	t.Run("InvalidPmQ", testInvalidPmQ)
	t.Run("InvalidPKNoneLinear", testInvalidPKNoneLinear)
	t.Run("InvalidPKT", testInvalidPKT)
	t.Run("InvalidPKOrder2", testInvalidPKOrder2)
	t.Run("InvalidPKFullOrder", testInvalidPKFullOrder)
	t.Run("ValidPQ", testValidPQ)
```

File: dh/sidh/internal/templates/curve_test.gotemp [hunk 13]
```diff
@@ -98,3 +102,316 @@ func BenchmarkThreePointLadder(b *testing.B) {
...
+
+// Left-to-right Montgomery ladder, Algorithm 4 in Costello-Smith
+// Input: ProjectivePoint P (xP, zP)
+// Output: x([scalar]P), z([scalar]P)
+func montgomeryLadder(cparams *ProjectiveCurveParameters, P *ProjectivePoint, scalar []uint8, random uint) ProjectivePoint {
+	var R0, R2, R1 ProjectivePoint
+	coefEq := CalcCurveParamsEquiv4(cparams) // for xDbl
+	aPlus2Over4 := CalcAplus2Over4(cparams)	 // for xDblAdd
+	R0 = *P					 // RO <- P
+	R1 = *P; xDbl(&R1, &coefEq)		 // R1 <- [2]P
```
Comment (armfazh) [inline dh/sidh/internal/templates/curve_test.gotemp:138]: Change this to `Pow2k(&R1, &coefEq, 1)` and remove xDbl above.

**Review by armfazh — CHANGES_REQUESTED** [review]
two minor changes needed so far

**Review by armfazh — APPROVED** [review]

File: dh/sidh/internal/templates/core.gotemp [hunk 14]
```diff
@@ -260,7 +261,7 @@ func DeriveSecretA(ss, prv []byte, pub3Pt *[3]Fp2) {
 }
 
 // Establishing shared keys in in 3-torsion group
-func DeriveSecretB(ss, prv []byte, pub3Pt *[3]Fp2) {
+func DeriveSecretB(ss, prv []byte, pub3Pt *[3]Fp2)  {
```
Comment (bwesterb) [inline dh/sidh/internal/templates/core.gotemp:264]: Run `go fmt`

**Review by bwesterb — APPROVED** [review]

**Pull request merged**